### PR TITLE
Shadow integration

### PR DESCRIPTION
**Pull request by AbbyGi**
This will add Shadow3 integration into sirepo-bluesky.
@mrakitin @jklynch

### CODE REVIEW

**Comment by mrakitin** [comment]
It's good we have the CI configured, so the API breaks were immediately captured. We will work on making the old API preserved as we add the support for new codes.

**Comment by mrakitin** [comment]
A working example:
```py
$ ipython
Python 3.8.8 (default, Apr 13 2021, 12:59:45)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.22.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: %run -i examples/prepare_det_env.py
   ...: import sirepo_bluesky.sirepo_detector as sd
   ...: import bluesky.plans as bp
   ...: sirepo_det = sd.SirepoDetector(sim_type='shadow', sim_id='NmYk68QS', sirepo_server='http://localhost:8000')
   ...: sirepo_det.select_optic('Aperture')
   ...: param1 = sirepo_det.create_parameter('horizontalSize')
   ...: param2 = sirepo_det.create_parameter('verticalSize')
   ...: sirepo_det.read_attrs = ['image', 'mean', 'photon_energy']
   ...: sirepo_det.configuration_attrs = ['horizontal_extent', 'vertical_extent', 'shape']
/Users/mrakitin/anaconda3/envs/sirepo_bluesky/lib/python3.8/site-packages/pims/image_reader.py:26: RuntimeWarning: PIMS image_reader.py could not find scikit-image. Falling back to matplotlib's imread(), which uses floats instead of integers. This may break your scripts.
(To ignore this warning, include the line "warnings.simplefilter("ignore", RuntimeWarning)" in your script.)
  warnings.warn(RuntimeWarning(ski_preferred))

In [2]: RE(bp.grid_scan([sirepo_det],
   ...:                 param1, 0, 1, 3,
   ...:                 param2, 0, 1, 2,
   ...:                 True))


Transient Scan ID: 1     Time: 2021-06-09 14:03:10
Persistent Unique Scan ID: '1d54b235-d7d3-4681-aae8-012978f69ae2'
New stream: 'primary'
+-----------+------------+--------------------------------+------------------------------+-----------------+
|   seq_num |       time | Aperture_sirepo_horizontalSize | Aperture_sirepo_verticalSize | sirepo_det_mean |
+-----------+------------+--------------------------------+------------------------------+-----------------+
|         1 | 14:03:23.5 |                          0.000 |                        0.000 |        9000.000 |
|         2 | 14:03:33.3 |                          0.000 |                        1.000 |        9000.000 |
|         3 | 14:03:41.6 |                          0.500 |                        1.000 |        9000.000 |
|         4 | 14:03:51.1 |                          0.500 |                        0.000 |        9000.000 |
|         5 | 14:04:00.5 |                          1.000 |                        0.000 |        9000.000 |
|         6 | 14:04:08.7 |                          1.000 |                        1.000 |        9000.000 |
+-----------+------------+--------------------------------+------------------------------+-----------------+
generator grid_scan ['1d54b235'] (scan num: 1)



Out[2]: ('1d54b235-d7d3-4681-aae8-012978f69ae2',)
```

File: sirepo_bluesky/shadow_handler.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+def read_shadow_file(filename, parameter=11):
+    """Read specified parameter from the Shadow3 output binary file.
+
+    Parameters
+    ----------
+    filename : str
+        Shadow3 output binary file (.dat)
+    parameter : int
+        The parameter to read (11=Energy [eV] by default)
```
Comment (mrakitin) [inline sirepo_bluesky/shadow_handler.py:13]: Consider using parameter 23 (https://github.com/oasys-kit/shadow3/blob/d007ae59a2038db4f9275f7bb026bd1b11549e5f/Shadow/ShadowTools.py#L68) for the total intensity.

**Comment by mrakitin** [comment]
With the updated handler, the data looks more realistic. Here is the code to run the grid scan:
```py
In [1]: %run -i examples/prepare_det_env.py
   ...: import sirepo_bluesky.sirepo_detector as sd
   ...: import bluesky.plans as bp
   ...: sirepo_det = sd.SirepoDetector(sim_type='shadow', sim_id='NmYk68QS', sirepo_server='http://localhost:8000')
   ...: sirepo_det.select_optic('Aperture')
   ...: param1 = sirepo_det.create_parameter('horizontalSize')
   ...: param2 = sirepo_det.create_parameter('verticalSize')
   ...: sirepo_det.read_attrs = ['image', 'mean', 'photon_energy', 'sirepo_json']
   ...: sirepo_det.configuration_attrs = ['horizontal_extent', 'vertical_extent', 'shape']

In [2]: RE(bp.grid_scan([sirepo_det],
   ...:                 param1, 0, 1, 11,
   ...:                 param2, 0, 1, 11,
   ...:                 True))
```

Live grid plot:
image

Get data for individual intensity distributions:
```py
In [8]: data = np.array(list(hdr.data('sirepo_det_image')))

In [9]: data.shape
Out[9]: (121, 100, 100)
```

Visualize a single intensity distribution:
```py
In [12]: plt.imshow(data[118])
```
image

cc @AbbyGi @jklynch @bnash

**Comment by mrakitin** [comment]
@AbbyGi, @jklynch, ready for your review!

**Review by jklynch — APPROVED** [review]

**Comment by mrakitin** [comment]
Thanks, @jklynch!

**Pull request merged**